storage/src/index/storage.rs

```diff
@@ -1,337 +1,465 @@
+                });
             }
-            Entry::Vacant(vacant_entry) => UpdateValueIterator {
-                next: None,
-                entry: Some(MapEntry::Vacant(vacant_entry)),
-                collisions_counter: &self.collisions,
-            },
         }
     }
 
-    /// Retrieve all values associated with a translated key, allowing for mutation and removal.
-    pub fn remove_iter(&mut self, key: &[u8]) -> RemoveValueIterator<T::Key, V> {
-        let translated_key = self.translator.transform(key);
-        let entry = self.map.entry(translated_key);
-        match entry {
-            Entry::Occupied(mut occupied_entry) => {
-                let record_ptr = occupied_entry.get_mut();
-                RemoveValueIterator {
-                    can_remove: false,
-                    prev_prev: None,
-                    prev: None,
-                    next: Some(record_ptr),
-                    entry: Some(occupied_entry),
-                    pruned_counter: &self.keys_pruned,
+    /// Insert a value at the given translated key, and prune any values that are no longer valid.
+    pub fn insert_and_prune(&mut self, key: &[u8], v: V, prune: impl Fn(&V) -> bool) {
+        let k = self.translator.transform(key);
+        match self.map.entry(k) {
+            Entry::Occupied(mut entry) => {
+                // Get entry
+                let entry = entry.get_mut();
+                let mut cursor = Cursor::new(entry, &self.collisions, &self.pruned);
+
+                // Remove anything that is prunable.
+                loop {
+                    let Some(old) = cursor.next() else {
+                        break;
+                    };
+                    if prune(old) {
+                        cursor.delete();
+                    }
                 }
+
+                // Add our new value.
+                cursor.insert(v);
+
+                // There must be something left, so we don't need to worry about checking
+                // for emptiness.
+            }
+            Entry::Vacant(entry) => {
+                // No collision, so we can just insert the value.
+                entry.insert(Record {
+                    value: v,
+                    next: None,
+                });
             }
-            Entry::Vacant(_) => RemoveValueIterator {
-                can_remove: false,
-                prev_prev: None,
-                prev: None,
-                next: None,
-                entry: None,
-                pruned_counter: &self.keys_pruned,
-            },
         }
     }
 
-    /// Remove all values associated with a translated key that match the `prune` predicate.
-    pub fn remove(&mut self, key: &[u8], prune: impl Fn(&V) -> bool) {
-        let mut iter = self.remove_iter(key);
-        while let Some(value) = iter.next() {
-            if prune(value) {
-                iter.remove();
+    /// Remove a value at the given translated key, and prune any values that are no longer valid.
```

Comment:
Maybe I'm not understanding the intent here, but the original comment seemed more accurate? (For example, it could be that no values are removed for the given translated key.)

Reply:
Good catch.